### PR TITLE
Neural network optimizers module

### PR DESCRIPTION
**Pull request by shretadas**
## Neural Network Optimizers Module

This PR adds a comprehensive neural network optimizers module implementing 5 standard optimization algorithms used in machine learning and deep learning.
Fixes #13662

### What's Added:

- Add SGD (Stochastic Gradient Descent) optimizer
- Add MomentumSGD with momentum acceleration
- Add NAG (Nesterov Accelerated Gradient) optimizer
- Add Adagrad with adaptive learning rates
- Add Adam optimizer combining momentum and RMSprop
- Include comprehensive doctests (61 tests, all passing)
- Add abstract BaseOptimizer for consistent interface
- Include detailed mathematical documentation
- Add educational examples and performance comparisons
- Follow repository guidelines: type hints, error handling, pure Python

Implements standard optimization algorithms for neural network training with educational focus and comprehensive testing coverage.

### Technical Details:

**Algorithms Implemented:**
- **SGD**: θ = θ - α∇θ (basic gradient descent)
- **MomentumSGD**: v = βv + (1-β)∇θ, θ = θ - αv
- **NAG**: Uses lookahead gradients for better convergence
- **Adagrad**: Adaptive learning rates per parameter  
- **Adam**: Combines momentum + adaptive learning rates

**Files Added:**

neural_network/optimizers/
├── _init_.py # Package initialization
├── README.md # Comprehensive documentation
├── base_optimizer.py # Abstract base class
├── sgd.py # Stochastic Gradient Descent
├── momentum_sgd.py # SGD with Momentum
├── nag.py # Nesterov Accelerated Gradient
├── adagrad.py # Adagrad optimizer
├── adam.py # Adam optimizer
├── test_optimizers.py # Comprehensive test suite
└── IMPLEMENTATION_SUMMARY.md # Technical implementation details

**Testing Coverage:**
- 61 comprehensive doctests (100% pass rate)
- Error handling for all edge cases
- Multi-dimensional parameter support
- Performance comparison examples

### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [x] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [x] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-13662".

### CODE REVIEW

File: neural_network/optimizers/adagrad.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+            parameters: Current parameter values
+            gradients: Gradients of loss function w.r.t. parameters
+            
+        Returns:
+            Updated parameters
+            
+        Raises:
+            ValueError: If parameters and gradients have different shapes
+        """
+        def _adagrad_update_recursive(params, grads, acc_grads):
```
[the remark this answers is not in the document]
Comment (shretadas) [reply neural_network/optimizers/adagrad.py:122]: def _check_and_update_recursive(
    params: list[float], 
    grads: list[float]
) -> list[float]:
    ...
